sympy/core/mul.py

```diff
@@ -2156,8 +2162,12 @@ def _keep_coeff(coeff, factors, clear=True, sign=False):
         return coeff
     if coeff is S.One:
         return factors
-    elif coeff is S.NegativeOne and not sign:
+    elif coeff.is_Float and equal_valued(coeff, S.One):
+        return Add(*[coeff*term for term in Add.make_args(factors)])
```

Comment:
```suggestion
        if (c,a:=factors.as_coeff_add())[0].is_Number and c:
            return coeff*c + a
        return return Add(*[coeff*term for term in Add.make_args(factors)])
```

Reply:
And perhaps similar for below. I don't see the need to store 1.0 as a coefficient of every term in an expression that looks like `x + y + 3`. Isn't `x + y + 3.0` sufficient?

Also, `as_coeff_add` should be modified consistent with `as_coeff_mul`, perhaps:
```python
>>> (x+1.).as_coeff_add()
(0, (1.00000000000000, x))
>>> (x+1.).as_coeff_Add()
(1.00000000000000, x)
```

I actually thought this would break things, however. I am surprised that it didn't.

Reply:
It did break some things. Hence the other changes.

It probably fixes lots of untested float-related bugs though. We don't make it easy for ourselves to handle floats in high-level routines by arbitrarily treating them differently in low-level ones.